### PR TITLE
SIM: parallelization

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

**Comment by peterrum** [comment]
This should work now. The PR used the new deal.II features `Utilities::MPI::RemotePointEvaluation` and `VectorTools::evaluate_at_points` :)

Next step is to replace the local marching-cube algorithms with the version from deal.II (https://github.com/dealii/dealii/pull/11828 - once merged).

File: include/adaflo/sharp_interface.h [hunk 1]
```diff
@@ -753,8 +761,10 @@ class MixedLevelSetSolver : public SharpInterfaceSolver
     euler_dofhandler.distribute_dofs(surface_fe_dim);
 
     euler_vector.reinit(euler_dofhandler.n_dofs());
+    euler_vector.update_ghost_values();
     VectorTools::
-      get_position_vector(euler_dofhandler, euler_vector, MappingQGeneric<dim - 1, dim>(4 /*TODO: this is a high number to well represent curved surfaces, the actual values is not that relevant*/));
+      get_position_vector(MappingQGeneric<dim - 1, dim>(4 /*TODO: this is a high number to well represent curved surfaces, the actual values is not that relevant*/), euler_dofhandler, euler_vector);
```
Comment (mschreter) [inline include/adaflo/sharp_interface.h:766]:
```suggestion
      get_position_vector(MappingQGeneric<dim - 1, dim>(4 /*TODO: this is a high number to well represent curved surfaces, the actual value is not that relevant*/), euler_dofhandler, euler_vector);
```

File: include/adaflo/sharp_interface.h [hunk 2]
```diff
@@ -851,6 +865,12 @@ class MixedLevelSetSolver : public SharpInterfaceSolver
                              navier_stokes_solver.get_dof_handler_u().get_fe().degree +
                                1);
 
+      navier_stokes_solver.solution.zero_out_ghost_values();
```
Comment (mschreter) [inline include/adaflo/sharp_interface.h:868]: This is something I don't know: Is it okay to `zero_out_ghost_values` before the output is actually written?

File: include/adaflo/sharp_interface_util.h [hunk 3]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
+  normal_solution.update_ghost_values();
+  curvature_solution.update_ghost_values();
 
-      cell_dim->get_dof_indices(local_dof_indices);
+  eval.template process_and_evaluate<T>(integration_values, buffer, fu);
 
-      constraints.distribute_local_to_global(buffer_dim, local_dof_indices, force_vector);
-    }
+  normal_solution.update_ghost_values();
+  curvature_solution.update_ghost_values();
```
Comment (mschreter) [inline include/adaflo/sharp_interface_util.h:935]: `zero_out_ghost_values`?

File: include/adaflo/sharp_interface_util.h [hunk 4]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
-                              surface_mapping,
-                              surface_fe,
-                              surface_quad,
-                              dof_handler.get_triangulation(),
-                              mapping);
-
-  // step 2) loop over all cells and evaluate curvature and normal in the cell-local
-  // quadrature points
-  //   and test with all test functions of the cell
+  using T = double;
```
Comment (mschreter) [inline include/adaflo/sharp_interface_util.h:776]: What is  the intention here?

File: tests/sharp_interfaces_01.cc [hunk 5]
```diff
@@ -378,24 +378,24 @@ test(const std::string &parameter_filename)
...
-                                            normal_vector_field,
-                                            curvature_solution,
-                                            force_vector_sharp_interface);
+  compute_force_vector_sharp_interface(surface_mesh,
+                                       surface_mapping,
+                                       surface_quad,
+                                       mapping,
+                                       dof_handler,
+                                       dof_handler_dim,
+                                       1.0,
```
Comment (mschreter) [inline tests/sharp_interfaces_01.cc:395]:
```suggestion
                                       1.0, /* TODO */
```

File: tests/sharp_interfaces_01.cc [hunk 6]
```diff
@@ -489,16 +489,16 @@ test(const std::string &parameter_filename)
...
-                                                level_set_solver.get_normal_vector(),
-                                                level_set_solver.get_curvature_vector(),
-                                                force_vector_sharp_interface);
+      compute_force_vector_sharp_interface(surface_mesh,
+                                           surface_mapping,
+                                           surface_quad,
+                                           mapping,
+                                           level_set_solver.get_dof_handler(),
+                                           level_set_solver.get_dof_handler_dim(),
+                                           1.0,
```
Comment (mschreter) [inline tests/sharp_interfaces_01.cc:498]:
```suggestion
                                           1.0, /*TODO*/
```

File: include/adaflo/sharp_interface_util.h [hunk 7]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
+        fe_eval.reinit(cell);
+
+        for (const auto q : fe_eval.quadrature_point_indices())
+          integration_values.push_back(fe_eval.JxW(q));
+      }
+
+    return integration_values;
+  }();
+
+  const auto fu = [&](const auto &values, const auto &cell_data) {
```
Comment (mschreter) [inline include/adaflo/sharp_interface_util.h:829]: maybe you can find a more verbose name than `fu` :-)

**Review by mschreter — APPROVED** [review]
Fine for me, some minor comments.

File: include/adaflo/sharp_interface.h (hunk 2 above)
Comment (peterrum) [reply include/adaflo/sharp_interface.h:868]: Yes, it is OK. `DataOut::build_patches()` does the actual work: it translates the vector data to a format that can be used for any output type. But I think the documentation is intentionally misleading^^

File: include/adaflo/sharp_interface_util.h (hunk 3 above)
Comment (peterrum) [reply include/adaflo/sharp_interface_util.h:935]: Indeed^^

File: include/adaflo/sharp_interface_util.h (hunk 4 above)
Comment (peterrum) [reply include/adaflo/sharp_interface_util.h:776]: I have added a comment.

File: include/adaflo/sharp_interface_util.h (hunk 7 above)
Comment (peterrum) [reply include/adaflo/sharp_interface_util.h:829]: Why does no one like the function name (see also: https://github.com/dealii/dealii/pull/11804#discussion_r595248782)? I guess you want to point out that I should instead use `foo`(and `bar`) ;) Instead, I went with `integration_function`.

**Comment by peterrum** [comment]
@mschreter Thanks for the feedback! I have applied the suggestions. Not your PR is coming!

File: include/adaflo/sharp_interface.h [hunk 8]
```diff
@@ -384,7 +384,13 @@ class LevelSetSolver
   void
   advance_concentration()
   {
+    velocity_solution.update_ghost_values();
+    velocity_solution_old.update_ghost_values();
+    velocity_solution_old_old.update_ghost_values();
     advection_operator->advance_concentration(this->time_stepping.step_size());
+    velocity_solution.zero_out_ghost_values();
```
Comment (kronbichler) [inline include/adaflo/sharp_interface.h:391]: This means that we are not backwards compatible with deal.II version 9.2 where we would have needed `zero_out_ghosts`. It probably won't matter too much because we have not paid attention to it elsewhere and there are no users on deal.II older than master right now (which was different in the past), but we would need to adjust the version here:
https://github.com/kronbichler/adaflo/blob/849a66c4c3afee3fb3a11eca54dec189edf2a807/CMakeLists.txt#L26
In other words, this demonstrates why I am hesitant to frequent interface changes in deal.II - it is a pain to support multiple versions.

File: include/adaflo/sharp_interface_util.h [hunk 9]
```diff
@@ -92,76 +61,58 @@ namespace dealii
...
+          for (const auto q : fe_eval.quadrature_point_indices())
+            {
+              const auto velocity = evaluation_values[counter++];
+
+              for (unsigned int comp = 0; comp < spacedim; ++comp)
+                {
+                  const auto i =
+                    euler_dofhandler.get_fe().component_to_system_index(comp, q);
+                  temp[i] = fe_eval.quadrature_point(q)[comp] + dt * velocity[comp];
+                }
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:115]: Nice use of the remote point evaluation.

File: include/adaflo/sharp_interface_util.h [hunk 10]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
-                              mapping);
-
-  // step 2) loop over all cells and evaluate curvature and normal in the cell-local
-  // quadrature points
-  //   and test with all test functions of the cell
+  using T = double; // type of data to be communicated (only |J|xW)
 
-  AffineConstraints<double> constraints; // TODO: use the right ones
+  const auto integration_points = [&]() {
+    std::vector<Point<spacedim>> integration_points;
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:779]: I suggest to simply fill `integration_points` inside a scope as a means of code structuring, rather than the lambda function. Or use a proper function that explicitly specifies the input arguments and not implicitly captures whole content of the surrounding class.

File: include/adaflo/sharp_interface_util.h [hunk 11]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
-      // gather normal
-      for (int i = 0; i < dim; ++i)
+        fe_eval.reinit(cell);
+
+        for (const auto q : fe_eval.quadrature_point_indices())
+          integration_values.push_back(fe_eval.JxW(q));
+      }
+
+    return integration_values;
+  }();
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:827]: Same here.

File: include/adaflo/sharp_interface_util.h [hunk 12]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
+          &dof_handler};
+
+        typename DoFHandler<spacedim>::active_cell_iterator cell_dim = {
+          &eval.get_triangulation(),
+          cell_data.cells[i].first,
+          cell_data.cells[i].second,
+          &dof_handler_dim};
+
+        const ArrayView<const Point<spacedim>> unit_points(
+          cell_data.reference_point_values.data() + cell_data.reference_point_ptrs[i],
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:857]: I realize that we might probably want to use a better name than `..._ptrs` for this in deal.II - this name does not really explain the CRS data structure in my opinion. We use `..._row_start` elsewhere IIRC.

File: include/adaflo/sharp_interface_util.h [hunk 13]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
+  const auto integration_function = [&](const auto &values, const auto &cell_data) {
+    AffineConstraints<double> constraints; // TODO: use the right ones
+
+    FEPointEvaluation<1, spacedim> phi_curvature(mapping, dof_handler.get_fe());
+
+    FESystem<spacedim>                    fe_dim(dof_handler.get_fe(), spacedim);
+    FEPointEvaluation<spacedim, spacedim> phi_normal(mapping, fe_dim);
+    FEPointEvaluation<spacedim, spacedim> phi_force(mapping, dof_handler_dim.get_fe());
+
+    std::vector<double>                  buffer;
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:838]: Question: Do you use this as
```suggestion
    std::vector<double>                  curvature_dof_values;
```

File: include/adaflo/sharp_interface_util.h [hunk 14]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
+    std::vector<double>                  buffer_dim;
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:839]:
```suggestion
    std::vector<double>                  normal_values;
```

File: include/adaflo/sharp_interface_util.h [hunk 15]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
 
-      // quadrature loop
-      for (unsigned int q = 0; q < n_points; ++q)
+          phi_normal.evaluate(cell,
+                              unit_points,
+                              make_array_view(buffer_dim),
+                              EvaluationFlags::values);
+        }
+
+        // perform operation on quadrature points
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:901]:
```suggestion
        // perform operation at quadrature points
```

File: include/adaflo/sharp_interface_util.h [hunk 16]
```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
...
+  curvature_solution.update_ghost_values();
 
-      cell_dim->get_dof_indices(local_dof_indices);
+  eval.template process_and_evaluate<T>(integration_values, buffer, integration_function);
 
-      constraints.distribute_local_to_global(buffer_dim, local_dof_indices, force_vector);
-    }
+  normal_solution.zero_out_ghost_values();
+  curvature_solution.zero_out_ghost_values();
+  force_vector.compress(VectorOperation::values::add);
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:936]: Not that it matters much, but I think the most common notation in deal.II is
```suggestion
  force_vector.compress(VectorOperation::add);
```

File: include/adaflo/sharp_interface_util.h [hunk 17]
```diff
@@ -941,9 +969,16 @@ compute_force_vector_sharp_interface(const Quadrature<dim - 1> &surface_quad,
   std::vector<double>                  buffer_dim;
```
Comment (kronbichler) [inline include/adaflo/sharp_interface_util.h:969]: While pre-existing, I think this variable would also be more self-explanatory by the context it is used for, which is the normal vector.

File: tests/sharp_interfaces_06.prm [hunk 18]
```diff
@@ -1,59 +1,34 @@
...
-  set output vtk files = 1
-end
+#mpirun : 2
+subsection Problem - specific set two - phase method =
+  mixed level set end subsection Two phase set density = 1. set density difference =
+    0. set viscosity = 0.1 set viscosity difference = 0 set surface tension =
+      1.0 set epsilon = 1.5 set gravity = 0. set concentration subdivisions =
+        3 set grad pressure compatible  = 1 set localize surface tension =
+          1 set curvature correction    = 1 set number reinit steps =
+            2 set number initial reinit steps =
```
Comment (kronbichler) [inline tests/sharp_interfaces_06.prm:24]: Can you check what happened here?

File: tests/sharp_interfaces_06.cc [hunk 19]
```diff
@@ -124,15 +124,18 @@ MicroFluidicProblem<dim>::run()
   navier_stokes_solver.setup_problem(Functions::ZeroFunction<dim>(dim));
   navier_stokes_solver.output_solution(parameters.output_filename);
 
-  Triangulation<dim - 1, dim> surface_mesh;
+  parallel::shared::Triangulation<dim - 1, dim> surface_mesh(mpi_communicator);
```
Comment (kronbichler) [inline tests/sharp_interfaces_06.cc:127]: Does this mean that we need to ensure deal.II is built with METIS? Or can we make the test optional? I would like to avoid a required dependency on too many external libraries (unless we can provide a spack recipe or something similar that helps the user along).

**Review by kronbichler — COMMENTED** [review]
This is nice progress. I have some comments on the structuring and variable names.

File: include/adaflo/sharp_interface.h (hunk 8 above)
Comment (peterrum) [reply include/adaflo/sharp_interface.h:391]: I have increased the version to 9.3. We are relying (and will be relying thanks to @mschreter) on so many new features from deal.II so that providing backwards compability will be quite hard to accomplish. What do you think?

File: include/adaflo/sharp_interface_util.h (hunk 10 above)
Comment (peterrum) [reply include/adaflo/sharp_interface_util.h:779]: Done.

File: include/adaflo/sharp_interface_util.h (hunk 11 above)
Comment (peterrum) [reply include/adaflo/sharp_interface_util.h:827]: Same here.

File: include/adaflo/sharp_interface_util.h (hunk 13 above)
Comment (peterrum) [reply include/adaflo/sharp_interface_util.h:838]: The problem here is that these vectors are reused: for block vectors buffer is also used to read the components of the normal vector...

File: include/adaflo/sharp_interface_util.h (hunk 14 above)
Comment (peterrum) [reply include/adaflo/sharp_interface_util.h:839]: ... and here the buffer is also used for the suface tension, which is why I would like to keep the names as they are.

File: tests/sharp_interfaces_06.cc (hunk 19 above)
Comment (peterrum) [reply tests/sharp_interfaces_06.cc:127]: As far as I know, if now METIS or Zoltan are available the class will use partitioning along a z-curve so it should work.

**Comment by peterrum** [comment]
@kronbichler Thanks for the review! I have applied the suggestions and have rebased the branch.

File: include/adaflo/sharp_interface.h (hunk 8 above)
Comment (kronbichler) [reply include/adaflo/sharp_interface.h:395]: I definitely agree, it is good to make the cut now due to the large new feature set. (I merely used it for demonstration of compatibility.)

File: tests/sharp_interfaces_06.cc (hunk 19 above)
Comment (kronbichler) [reply tests/sharp_interfaces_06.cc:127]: Ah ok, that would be great. So let's keep it like this (I won't test it immediately because I have metis on my installations.)

**Comment by kronbichler** [comment]
I will give it a final look a bit later.

**Review by kronbichler — APPROVED** [review]
